test/gtest/ucp/test_ucp_reconfigure.cc

```diff
@@ -0,0 +1,264 @@
+            m_uct_eps.push_back(uct_ep);
+        }
+    }
+
+    m_cfg_index = ep()->cfg_index;
+}
+
+ucp_tl_bitmap_t test_ucp_reconfigure::entity::ep_tl_bitmap() const
+{
+    ucp_tl_bitmap_t tl_bitmap = UCS_STATIC_BITMAP_ZERO_INITIALIZER;
+
+    if (ep() == NULL) {
+        return tl_bitmap;
+    }
+
+    for (ucp_lane_index_t lane = 0; lane < ucp_ep_num_lanes(ep()); ++lane) {
+        UCS_STATIC_BITMAP_SET(&tl_bitmap, ucp_ep_get_rsc_index(ep(), lane));
+    }
+
+    return tl_bitmap;
+}
+
+void test_ucp_reconfigure::entity::connect(const entity &other,
+                                           bool is_exclude_iface)
+{
+    auto worker_addr = other.get_address(false);
+    ucp_tl_bitmap_t tl_bitmap;
+    unsigned addr_indices[UCP_MAX_LANES];
+
+    tl_bitmap = is_exclude_iface ? UCS_STATIC_BITMAP_NOT(other.ep_tl_bitmap()) :
+                                   ucp_tl_bitmap_max;
+
+    UCS_ASYNC_BLOCK(&worker()->async);
+    ASSERT_UCS_OK(ucp_ep_create_to_worker_addr(worker(), &tl_bitmap,
+                                               &worker_addr->second,
+                                               UCP_EP_INIT_CREATE_AM_LANE,
+                                               "reconfigure test", addr_indices,
+                                               &m_ep));
+
+    m_ep->conn_sn = 0;
+    ASSERT_TRUE(ucp_ep_match_insert(worker(), m_ep, worker_addr->second.uuid,
+                                    m_ep->conn_sn, UCS_CONN_MATCH_QUEUE_EXP));
+
+    ASSERT_UCS_OK(ucp_wireup_send_request(m_ep));
+    UCS_ASYNC_UNBLOCK(&worker()->async);
+
+    store_config();
+}
+
+bool test_ucp_reconfigure::entity::has_matching_lane(ucp_ep_h ep,
+                                                     ucp_lane_index_t lane_idx,
+                                                     const entity &other) const
+{
+    const auto lane     = &ucp_ep_config(ep)->key.lanes[lane_idx];
+    auto context        = m_worker_entity->ucph();
+    const auto resource = &context->tl_rscs[ucp_ep_get_rsc_index(ep, lane_idx)];
+    auto addr           = other.get_address(true);
+    const ucp_address_entry_t *ae;
+
+    ucs_carray_for_each(ae, addr->second.address_list,
+                        addr->second.address_count) {
+        if ((resource->tl_name_csum == ae->tl_name_csum) &&
+            ucp_wireup_is_lane_connected(ep, lane_idx, ae)) {
+            EXPECT_EQ(ae->sys_dev, lane->dst_sys_dev);
+            EXPECT_EQ(ae->md_index, lane->dst_md_index);
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void test_ucp_reconfigure::entity::verify_configuration(
+        const entity &other) const
+{
+    unsigned reused                  = 0;
+    const ucp_lane_index_t num_lanes = ucp_ep_num_lanes(ep());
+
+    EXPECT_EQ(num_lanes, ucp_ep_num_lanes(other.ep()));
+
+    for (ucp_lane_index_t lane = 0; lane < num_lanes; ++lane) {
+        /* Verify local and remote lanes are identical */
+        EXPECT_TRUE(has_matching_lane(ep(), lane, other));

```

Comment:
maybe return "matching lane" instead of `bool`? then add it into std::set which can test by size, uniq and values

Reply:
std::set will not detect repetitions

Reply:
exactly, `std::set` cannot contain duplicates, why is it possible? should not all lanes be unique?

Reply:
they are not unique because of paths

Reply:
then how do you control if this is a re-used lane or newly created but with the same transport?

Reply:
In wireup.c we count how many times we are using the same address entry and expect it will be max_paths.
In test_reconfigure I cannot keep counter per address entry in std::set.

Reply:
It can be done with std::map or std::pair but it's a bit too complicated for now IMO